gapis/resolve/dependencygraph2/graph_visualization/graph_visualization.go

```diff
@@ -19,33 +19,62 @@ import (
 	"fmt"
 	"github.com/google/gapid/gapis/api"
 	"github.com/google/gapid/gapis/resolve/dependencygraph2"
+	"github.com/google/gapid/gapis/service"
 	"github.com/google/gapid/gapis/service/path"
 )
 
+const (
+	EMPTY = "Empty"
+)
+
 func createGraphFromDependencyGraph(dependencyGraph dependencygraph2.DependencyGraph) (*graph, error) {
 
 	currentGraph := createGraph(0)
-	currentHierarchy := &api.Hierarchy{}
+	idToBuilder := map[api.ID]api.GraphVisualizationBuilder{}
 
 	err := dependencyGraph.ForeachNode(
 		func(nodeId dependencygraph2.NodeID, dependencyNode dependencygraph2.Node) error {
 			if cmdNode, ok := dependencyNode.(dependencygraph2.CmdNode); ok {
 				cmdNodeId := cmdNode.Index[0]
 				command := dependencyGraph.GetCommand(api.CmdID(cmdNodeId))
-				commandName := command.CmdName()
+				isReal := api.CmdID(cmdNodeId).IsReal()
+				if !isReal {
+					return nil
+				}
 
 				if graphVisualizationAPI, ok := command.API().(api.GraphVisualizationAPI); ok {
-					label := graphVisualizationAPI.GetCommandLabel(currentHierarchy, command)
-					label += fmt.Sprintf("%s%d", commandName, cmdNodeId)
-					label += graphVisualizationAPI.GetSubCommandLabel(cmdNode.Index)
+					if _, ok := idToBuilder[command.API().ID()]; !ok {
+						idToBuilder[command.API().ID()] = graphVisualizationAPI.GetGraphVisualizationBuilder()
+					}
+					builder := idToBuilder[command.API().ID()]
+
+					label, name := "", ""
+					if len(cmdNode.Index) == 1 {
+						label = builder.GetCommandLabel(command, cmdNodeId)
+						name = command.CmdName()
+					} else if len(cmdNode.Index) > 1 {
+						commandName := fmt.Sprintf("%s_%d", command.CmdName(), cmdNodeId)
+						dependencyNodeAccesses := dependencyGraph.GetNodeAccesses(nodeId)
+						subCommandName := EMPTY
+						if len(dependencyNodeAccesses.InitCmdNodes) > 0 {
+							subCmdDependencyNode := dependencyGraph.GetNode(dependencyNodeAccesses.InitCmdNodes[0])
+							if subCmdNode , ok := subCmdDependencyNode.(dependencygraph2.CmdNode) ; ok {
+								subCmdId := subCmdNode.Index[0]
```

Comment:
`subCmdNode.Index` *should* always have exactly one element, but it is still probably worth checking that, rather than potentially segfaulting.

Reply:
Considering that all previous cmdNode.Index have been checked to contains exactly one element, I think for this case is safe,, but for unexpected segfaulting I am adding the checking anyway.